### PR TITLE
[Autotune Inductor] Some clean up and dataclassing

### PR DESCRIPTION
**Pull request by drisspg**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139157

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/select_algorithm.py [hunk 1]
```diff
@@ -76,6 +77,55 @@ class KernelNamespace:
...
+
+    @classmethod
+    def from_choice_args(
+        cls,
+        example_inputs: List[torch.Tensor],
+        example_inputs_extern: List[torch.Tensor],
+        out: torch.Tensor,
+        out_extern: torch.Tensor,
+        expected: Optional[torch.Tensor] = None,
+    ) -> "AutotuneArgs":
```
Comment (Skylion007) [inline torch/_inductor/select_algorithm.py:116]:
```suggestion
    ) -> Self:
```
`from typing_extensions import Self`

File: torch/_inductor/select_algorithm.py [hunk 2]
```diff
@@ -76,6 +77,55 @@ class KernelNamespace:
...
+@dataclasses.dataclass
+class BenchmarkTensors:
+    """Represents a set of inputs and outputs for autotuining with a template"""
+
+    input_tensors: List[torch.Tensor]
+    output_tensor: Optional[torch.Tensor]
+
+    def unpack(self):
+        return dataclasses.astuple(self)
+
```
Comment (Skylion007) [inline torch/_inductor/select_algorithm.py:89]:
```suggestion
_T = TypeVar("_T", bound="AutotuneArgs")
```

File: torch/_inductor/select_algorithm.py [hunk 3]
```diff
@@ -76,6 +77,55 @@ class KernelNamespace:
...
+        out: torch.Tensor,
+        out_extern: torch.Tensor,
+        expected: Optional[torch.Tensor] = None,
+    ) -> "AutotuneArgs":
+        """Factory method to create AutotuneInputs from separate inputs/outputs"""
+        return cls(
+            triton=BenchmarkTensors(example_inputs, out),
+            extern=BenchmarkTensors(example_inputs_extern, out_extern),
+            expected=expected,
+        )
```
Comment (Skylion007) [inline torch/_inductor/select_algorithm.py:122]:
```suggestion
    @classmethod
    def from_choice_args(
        cls: Type[_T],  # Typing cls to be the class itself, allowing subclasses
        example_inputs: List[torch.Tensor],
        example_inputs_extern: List[torch.Tensor],
        out: torch.Tensor,
        out_extern: torch.Tensor,
        expected: Optional[torch.Tensor] = None,
    ) -> _T:  # Return type matches the type of the class calling it
        """Factory method to create AutotuneInputs from separate inputs/outputs"""
        return cls(
            triton=BenchmarkTensors(example_inputs, out),
            extern=BenchmarkTensors(example_inputs_extern, out_extern),
            expected=expected,
        )

Comment (Skylion007) [reply torch/_inductor/select_algorithm.py:122]: Otherwise, the typing will be incorrect for subclasses.

**Review by Skylion007 — DISMISSED** [review]

**Review by eellison — COMMENTED** [review]
re-request review when tests passing plz

File: torch/_inductor/select_algorithm.py [hunk 4]
```diff
@@ -76,6 +77,59 @@ class KernelNamespace:
 extern_kernels = KernelNamespace()
 
 
+_T = TypeVar("_T", bound="AutotuneArgs")
+
+
+@dataclasses.dataclass
+class BenchmarkTensors:
+    """Represents a set of inputs and outputs for autotuining with a template"""
```
Comment (eellison) [inline torch/_inductor/select_algorithm.py:85]: nit: /s/autotuining/autotuning

File: torch/_inductor/select_algorithm.py [hunk 5]
```diff
@@ -76,6 +77,59 @@ class KernelNamespace:
...
+    output_tensor: Optional[torch.Tensor]
+
+    def unpack(self):
+        return self.input_tensors, self.output_tensor
+
+
+@dataclasses.dataclass
+class AutotuneArgs:
+    """During autotuning, we need to pass the same inputs to all choices.
+    Note:
```
Comment (eellison) [inline torch/_inductor/select_algorithm.py:97]: Nit: they are the same buffers, but with the offset either included or not. The comment below explains why:
```
            if isinstance(choice, ExternKernelCaller):
                # aten kernels want the offset baked in for sliced tensors
                result = choice.benchmark(*example_inputs_extern, out=out_extern)
            else:
                # triton templates want the base pointer for sliced tensors
```

Can we retain this comment somewhere in the code ?

**Review by eellison — APPROVED** [review]
looks good!

**Pull request closed without merging**